### PR TITLE
Publish snapshots to maven via GHA

### LINKED ISSUE #361: Publish snapshots to maven via GHA
Coming from [opensearch-build#3185](https://github.com/opensearch-project/opensearch-build/issues/3185)

### What is happening?

We are de-coupling the task of publishing the maven snapshots from centralized build workflow to individual repositories. What this means is each repo can now publish maven snapshots using GitHub Actions.
This change unblocks the dependent components from waiting for a successful build before they can consume the snapshots. This also ensures that all snapshots are independent and up to date.

### What do you need to do?
1. Determine what artifacts you publish to maven. Most plugins publish only zips. [Example](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/).
2. Add below repository under `publishing` in build.gradle:

```
    repositories {
        maven {
            name = "Snapshots"
            url = "https://aws.oss.sonatype.org/content/repositories/snapshots"
            credentials {
                username "$System.env.SONATYPE_USERNAME"
                password "$System.env.SONATYPE_PASSWORD"
            }
        }
    }
```
[Example](https://github.com/opensearch-project/security/blob/main/build.gradle#L233-L242)

3. Determine what gradle task will publish the content to snapshots repo. Example: `./gradlew publishPluginZipPublicationToSnapshotsRepository`.
4. Create a pull request to set up a GitHub action that will fetch the credentials and publish the snpahots to https://aws.oss.sonatype.org/ [See sample workflow](https://github.com/opensearch-project/security/blob/main/.github/workflows/maven-publish.yml).
5. Add @gaiksaya as a reviewer, who will also take care of creating the IAM role that is required to fetch the credentials and add it to the GitHub Secrets.
### Testing

Test by editing the snapshot repository by updating the Snapshots target repository from maven to a local file system path.

build.gradle
```
        maven {
            name = "Snapshots" //  optional target repository name
            url = "https://aws.oss.sonatype.org/content/repositories/snapshots"
            credentials {
                username "$System.env.SONATYPE_USERNAME"
                password "$System.env.SONATYPE_PASSWORD"
            }
        }
```
->
```
        maven {
            name = "Snapshots" //  optional target repository name
            url = "snapshots"
        }
```
Run the command ./gradlew <taskName> which will publish the files in the file system under the 'snapshots' folder, then scan with `find` or your favor directory scanner to see the zip + pom with sha files

```
% find snapshots | sort                                                                                                                                                                               ~/git/security
snapshots
snapshots/org
snapshots/org/opensearch
snapshots/org/opensearch/plugin
snapshots/org/opensearch/plugin/opensearch-security
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha512
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha512
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha512
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.md5
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha1
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha256
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha512
```

Please feel free to reach out to @opensearch-project/engineering-effectiveness.

### PR DESCRIPTION
**Pull request by lezzago**
### Description
Publish snapshots to maven via GHA
 
### Issues Resolved
#361
 
### Tests
```
find snapshots/ | sort 
snapshots/
snapshots//org
snapshots//org/opensearch
snapshots//org/opensearch/common-utils
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-javadoc.jar
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-javadoc.jar.md5
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-javadoc.jar.sha1
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-javadoc.jar.sha256
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-javadoc.jar.sha512
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-sources.jar
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-sources.jar.md5
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-sources.jar.sha1
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-sources.jar.sha256
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1-sources.jar.sha512
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.jar
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.jar.md5
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.jar.sha1
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.jar.sha256
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.jar.sha512
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.pom
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.pom.md5
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.pom.sha1
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.pom.sha256
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/common-utils-3.0.0.0-20230222.162217-1.pom.sha512
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/maven-metadata.xml
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/maven-metadata.xml.md5
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha1
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha256
snapshots//org/opensearch/common-utils/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha512
snapshots//org/opensearch/common-utils/maven-metadata.xml
snapshots//org/opensearch/common-utils/maven-metadata.xml.md5
snapshots//org/opensearch/common-utils/maven-metadata.xml.sha1
snapshots//org/opensearch/common-utils/maven-metadata.xml.sha256
snapshots//org/opensearch/common-utils/maven-metadata.xml.sha512
```

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/maven-publish.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        with:
+          role-to-assume: ${{ secrets.PUBLISH_SNAPSHOTS_ROLE }}
+          aws-region: us-east-1
+      - name: publish snapshots to maven
+        run: |
+          export SONATYPE_USERNAME=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-username --query SecretString --output text)
+          export SONATYPE_PASSWORD=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-password --query SecretString --output text)
+          echo "::add-mask::$SONATYPE_USERNAME"
+          echo "::add-mask::$SONATYPE_PASSWORD"
+          ./gradlew publishPluginZipPublicationToSnapshotsRepository
```
Comment (gaiksaya) [inline .github/workflows/maven-publish.yml:39]: I believe the task should be something else right? Since we are not publishing zips here

Comment (lezzago) [reply .github/workflows/maven-publish.yml:39]: good call

Comment (lezzago) [reply .github/workflows/maven-publish.yml]: fixed now

**Comment by gaiksaya** [comment]
Something does not seem right https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/1.3.8.0-SNAPSHOT/
https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/3.0.0.0-SNAPSHOT/

The above URLs do not contain javadoc and sources jars whereas your output does.

**Comment by lezzago** [comment]
> Something does not seem right https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/1.3.8.0-SNAPSHOT/ https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/3.0.0.0-SNAPSHOT/
> 
> The above URLs do not contain javadoc and sources jars whereas your output does.

This is due to breaking changes just introduced to core, so the main branch wont succeed in its build right now. Used a different version and tested the change.

**Comment by gaiksaya** [comment]
> > Something does not seem right https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/1.3.8.0-SNAPSHOT/ https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/3.0.0.0-SNAPSHOT/
> > The above URLs do not contain javadoc and sources jars whereas your output does.
> 
> This is due to breaking changes just introduced to core, so the main branch wont succeed in its build right now. Used a different version and tested the change.

Yeah does not matter which version the artifacts should be the same. @prudhvigodithi any idea?

**Comment by AWSHurneyt** [comment]
> > > Something does not seem right https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/1.3.8.0-SNAPSHOT/ https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/3.0.0.0-SNAPSHOT/
> > > The above URLs do not contain javadoc and sources jars whereas your output does.
> > 
> > 
> > This is due to breaking changes just introduced to core, so the main branch wont succeed in its build right now. Used a different version and tested the change.
> 
> Yeah does not matter which version the artifacts should be the same. @prudhvigodithi any idea?

Per offline discussion with @gaiksaya, removing the `backport 2.x` tag from this PR as these changes have been manually cherry-picked to the `2.x` branch in PR https://github.com/opensearch-project/common-utils/pull/365 so that this PR can remain open to continue the discussion above.

**Comment by lezzago** [comment]
> > > Something does not seem right https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/1.3.8.0-SNAPSHOT/ https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/common-utils/3.0.0.0-SNAPSHOT/
> > > The above URLs do not contain javadoc and sources jars whereas your output does.
> > 
> > 
> > This is due to breaking changes just introduced to core, so the main branch wont succeed in its build right now. Used a different version and tested the change.
> 
> Yeah does not matter which version the artifacts should be the same. @prudhvigodithi any idea?

We resolved the common-utils issues. Can we get approval now?

**Review by AWSHurneyt — APPROVED** [review]

**Review by gaiksaya — APPROVED** [review]

**Pull request merged**